ee/api/role.py

```diff
@@ -0,0 +1,110 @@
+from typing import cast
+
+from django.db import IntegrityError
+from rest_framework import mixins, serializers, viewsets
+from rest_framework.permissions import SAFE_METHODS, BasePermission, IsAuthenticated
+
+from ee.models.role import Role, RoleMembership
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.api.shared import UserBasicSerializer
+from posthog.models import OrganizationMembership
+from posthog.models.user import User
+from posthog.permissions import OrganizationMemberPermissions, extract_organization
+
+DEFAULT_ROLE_NAME = "Write"
+
+
+class RolePermissions(BasePermission):
+    """
+    Requires organization admin level to change object, allows everyone read
+    """
+
+    message = "You need to have admin level or higher."
+
+    def has_permission(self, request, view):
+        organization = extract_organization(request.user)
+
+        requesting_membership: OrganizationMembership = OrganizationMembership.objects.get(
+            user_id=cast(User, request.user).id,
+            organization=organization,
+        )
+
+        if request.method in SAFE_METHODS or requesting_membership.level >= OrganizationMembership.Level.ADMIN:
+            return True
+        return False
+
+
+class RoleSerializer(serializers.ModelSerializer):
+    created_by = UserBasicSerializer(read_only=True)
+
+    class Meta:
+        model = Role
+        fields = ["id", "name", "created_at", "created_by"]
+        read_only_fields = ["id", "created_at", "created_by"]
+
+    def validate_name(self, name):
+        if Role.objects.filter(name__iexact=name).exists():
+            raise serializers.ValidationError("There is already a role with this name.", code="unique")
+        return name
+
+    def create(self, validated_data):
+        validated_data["organization"] = self.context["request"].user.organization
+        return super().create(validated_data)
+
+
+class RoleViewSet(
+    StructuredViewSetMixin,
+    mixins.ListModelMixin,
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    mixins.UpdateModelMixin,
+    mixins.DestroyModelMixin,
+    viewsets.GenericViewSet,
+):
+    permission_classes = [
+        IsAuthenticated,
+        OrganizationMemberPermissions,
+        RolePermissions,
+    ]
+    serializer_class = RoleSerializer
+    queryset = Role.objects.all()
+
+
+class RoleMembershipSerializer(serializers.ModelSerializer):
+    user = UserBasicSerializer(read_only=True)
+    role_id = serializers.UUIDField(read_only=True)
+
+    user_uuid = serializers.UUIDField(required=True, write_only=True)
```

Comment:
We should probably allow bulk updating here, as I imagine for big teams, it will be verrryyy annoying if they have to add members one by one to a role.

Reply:
Bulk create/updates on the API is not that intuitive here as you can see with the `.bulk_create` I've been testing with, I see us mostly doing it from the frontend, check the API calls + logic for `dashboard_collaborators` for something similar

Reply:
Will delete the commented out code below for bulk creation in another commit unless there's valid reasons why we want to try to make bulk creation work on the API end

Reply:
Works